editoast/src/modelsv2/connection_pool/db_connection_pool.rs

```diff
@@ -0,0 +1,85 @@
+use diesel_async::pooled_connection::deadpool::Object;
+use diesel_async::pooled_connection::deadpool::Pool;
+#[cfg(test)]
+use diesel_async::pooled_connection::AsyncDieselConnectionManager;
+#[cfg(test)]
+use diesel_async::AsyncConnection;
+use diesel_async::AsyncPgConnection;
+use std::sync::Arc;
+#[cfg(test)]
+use tokio::sync::OwnedRwLockWriteGuard;
+#[cfg(test)]
+use tokio::sync::RwLock;
+#[cfg(test)]
+use url::Url;
+
+use super::DbConnectionError;
+
+#[derive(Clone)]
+pub struct DbConnectionPoolV2 {
+    pub pool: Arc<Pool<AsyncPgConnection>>,
+    #[cfg(test)]
+    pub test_connection: Option<Arc<RwLock<Object<AsyncPgConnection>>>>,
+}
+
+impl DbConnectionPoolV2 {
+    pub fn new(pool: Pool<AsyncPgConnection>) -> Self {
+        Self {
+            pool: Arc::new(pool),
+            #[cfg(all(test))]
+            test_connection: None,
+        }
+    }
+
+    #[cfg(test)]
+    pub async fn new_for_tests(
+        pool: Arc<Pool<AsyncPgConnection>>,
+    ) -> Result<Self, DbConnectionError> {
+        let mut conn = pool.get().await.map_err(DbConnectionError::DeadpoolPool)?;
+        conn.begin_test_transaction()
+            .await
+            .map_err(DbConnectionError::DieselError)?;
+        let test_connection = RwLock::new(conn);
+        let test_connection = Arc::new(test_connection);
+
+        Ok(Self {
+            pool,
+
+            test_connection: Some(test_connection),
+        })
+    }
+
+    #[cfg(test)]
+    pub async fn get(
+        &self,
+    ) -> Result<OwnedRwLockWriteGuard<Object<AsyncPgConnection>>, DbConnectionError> {
+        if let Some(test_connection) = &self.test_connection {
+            let connection = test_connection.clone().write_owned().await;
+            Ok(connection)
+        } else {
+            Err(DbConnectionError::TestConnection)
+        }
+    }
+
+    #[cfg(not(test))]
```

Comment:
Maybe to avoid so many feature-gated functions, you could do only 3 parts like the following.

```rust
impl DbConnectionPoolV2 {
    pub fn new() -> Self {}
}

#[cfg(not(test))]
impl DbConnectionPoolV2 {
    pub async fn get() {}
}

#[cfg(test)]
impl DbConnectionPoolV2 {
    pub async fn new_for_tests() {}
    pub async fn get() {}
}
```

Note that it might become less important if some of the functions disappear.